### PR TITLE
adding authorizer for the api export virtual workspace

### LINKED ISSUE #1151: Add authorization checks (SAR) for APIExport virtual workspace
We want to make sure that the user's access the Virtual Workspace for an APIExport, that they have permissions.

I believe that that this means the user has access to the sub resource '/content' on the given workspace type being used.

It sounds like this to me, is the check that we will run for every interaction with a given virtual workspace.

### PR DESCRIPTION
**Pull request by shawn-hurley**
## Summary
Adding the authorizer for apiexport that will check that the requesting user has access to the apiexport context subresource for the specific API export. the verb being used will be forwarded along to be checked.

example: watch cowboys

request path: `/services/apiexport/root:org:ws/cowboys-export/clusters/*/api/v1/cowboys`

SAR check would ask if I can `list` on the specific `cowboys-export` `/content` subresource.

## Related issue(s)
Follow on for: https://github.com/kcp-dev/kcp/pull/1176
Fixes #1151

### CODE REVIEW

File: pkg/virtual/apiexport/builder/build.go [hunk 1]
```diff
@@ -177,6 +180,34 @@ func BuildVirtualWorkspace(
...
+
+		apiExportCluster, apiExportName := parts[0], parts[1]
+		authz, err := delegated.NewDelegatedAuthorizer(logicalcluster.New(apiExportCluster), client)
+		if err != nil {
+			return authorizer.DecisionNoOpinion, "error", err
+		}
+
+		SARAttributes := authorizer.AttributesRecord{
+			User:       attr.GetUser(),
+			Verb:       "content",
```
Comment (sttts) [inline pkg/virtual/apiexport/builder/build.go]: subresource, not content

File: pkg/virtual/apiexport/builder/build.go [hunk 2]
```diff
@@ -199,15 +193,18 @@ func getAuthorizer(client kubernetes.ClusterInterface) authorizer.AuthorizerFunc
...
-			APIVersion: apisv1alpha1.SchemeGroupVersion.Version,
-			Resource:   "apiexport",
+			APIGroup:        apisv1alpha1.SchemeGroupVersion.Group,
+			APIVersion:      apisv1alpha1.SchemeGroupVersion.Version,
+			User:            attr.GetUser(),
+			Verb:            attr.GetVerb(),
+			Name:            apiExportName,
+			Resource:        "apiexports",
+			ResourceRequest: true,
+			Subresource:     "content",
```
Comment (sttts) [inline pkg/virtual/apiexport/builder/build.go:203]: see comment about subresource vs. verb in the other commit

File: test/e2e/virtual/apiexport/virtualworkspace_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	require.NoError(t, err, "failed to construct wildwest cluster client for server")
+
+	framework.AdmitWorkspaceAccess(t, ctx, kubeClusterClient, orgClusterName, []string{"user-1"}, nil, []string{"member"})
+	framework.AdmitWorkspaceAccess(t, ctx, kubeClusterClient, serviceProviderWorkspace, []string{"user-1"}, nil, []string{"member", "access"})
+	//Create provider
+	setUpServiceProvider(ctx, dynamicClients, kcpClients, kubeClusterClient, serviceProviderWorkspace, t)
+	// Need to Create a consumer
+	bindConsumerToProvider(ctx, consumerWorkspace, serviceProviderWorkspace, t, kcpClients)
+	// Need to create a cowby in Consumer
+	createCowboyInConsumer(ctx, t, consumerWorkspace, wildwestClusterClient)
```
Comment (sttts) [inline test/e2e/virtual/apiexport/virtualworkspace_test.go:79]: these comments are not helpful

File: test/e2e/virtual/apiexport/virtualworkspace_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	require.NoError(t, err)
+	require.Equal(t, 1, len(cowboysProjected.Items))
+
+	// Attempt to use VW using user-1 should expect an error
+	t.Logf("Make sure that user-1 is denied")
+	user1VC := virtualClusterClient{config: userConfig("user-1", cfg)}
+	wwUser1VC, err := user1VC.Cluster(serviceProviderWorkspace, "today-cowboys")
+	require.NoError(t, err)
+	_, err = wwUser1VC.Cluster(logicalcluster.Wildcard).WildwestV1alpha1().Cowboys("").List(ctx, metav1.ListOptions{})
+	require.Error(t, err)
```
Comment (sttts) [inline test/e2e/virtual/apiexport/virtualworkspace_test.go:95]: check that it is forbidden error

**Comment by sttts** [comment]
Some small comments.

File: test/e2e/virtual/apiexport/virtualworkspace_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	t.Logf("Get Cowboys with user-1")
+	require.Eventually(t, func() bool {
+		cb, err := wwUser1VC.Cluster(logicalcluster.Wildcard).WildwestV1alpha1().Cowboys("").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return false
+		}
+		require.Equal(t, 1, len(cb.Items))
+		return true
+	}, wait.ForeverTestTimeout, time.Millisecond*100, "expected user-1 to list cowboys from virtual workspace")
+
```
Comment (sttts) [inline test/e2e/virtual/apiexport/virtualworkspace_test.go:115]: would like to see a case where a user or service account gets read-only access.

File: pkg/virtual/apiexport/builder/build.go (hunk 2 above)
Comment (sttts) [reply pkg/virtual/apiexport/builder/build.go:203]: correct now. Github has issues it seems.

File: test/e2e/virtual/apiexport/virtualworkspace_test.go (hunk 4 above)
Comment (shawn-hurley) [reply test/e2e/virtual/apiexport/virtualworkspace_test.go]: completed

File: test/e2e/virtual/apiexport/virtualworkspace_test.go (hunk 5 above)
Comment (shawn-hurley) [reply test/e2e/virtual/apiexport/virtualworkspace_test.go:124]: Added a test that checks that a user can only have `List` and not allowed to update

File: pkg/virtual/apiexport/builder/build.go [hunk 6]
```diff
@@ -177,6 +174,36 @@ func BuildVirtualWorkspace(
 
 			return apiReconciler, nil
 		},
+		Authorizer: getAuthorizer(kubeClusterClient),
+	}
+}
+
+func getAuthorizer(client kubernetes.ClusterInterface) authorizer.AuthorizerFunc {
```
Comment (sttts) [inline pkg/virtual/apiexport/builder/build.go:181]: https://en.wikipedia.org/wiki/Lambda_calculus#%CE%B7-reduction

**Pull request merged**